File: odoo-client.js
/**
 * odoo-client.js
 * Handles XML-RPC communication with Odoo instance
 */

class OdooClient {
    constructor() {
        this.uid = null;
    }

    getConfig() {
        return window.AppConfig.getAll();
    }

    /**
     * Helper to construct XML-RPC methodCall
     */
    buildXmlCall(methodName, params) {
        let paramsXml = '';
        params.forEach(p => {
            paramsXml += `<param><value>${this.toXmlRpcValue(p)}</value></param>`;
        });

        return `<?xml version="1.0"?>
<methodCall>
    <methodName>${methodName}</methodName>
    <params>${paramsXml}</params>
</methodCall>`;
    }

    /**
     * Convert JS value to XML-RPC value
     */
    toXmlRpcValue(val) {
        if (val === null || val === undefined) return '<nil/>';

        const type = typeof val;

        if (type === 'boolean') return `<boolean>${val ? '1' : '0'}</boolean>`;
        if (type === 'number') {
            return Number.isInteger(val)
                ? `<int>${val}</int>`
                : `<double>${val}</double>`;
        }
        if (type === 'string') return `<string>${this.escapeXml(val)}</string>`;

        if (Array.isArray(val)) {
            const items = val.map(v => `<value>${this.toXmlRpcValue(v)}</value>`).join('');
            return `<array><data>${items}</data></array>`;
        }

        if (type === 'object') {
            const members = Object.keys(val).map(k => `
                <member>
                    <name>${this.escapeXml(k)}</name>
                    <value>${this.toXmlRpcValue(val[k])}</value>
                </member>
            `).join('');
            return `<struct>${members}</struct>`;
        }

        return `<string>${this.escapeXml(String(val))}</string>`;
    }

    escapeXml(str) {
        return str.replace(/&/g, '&amp;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;')
            .replace(/"/g, '&quot;')
            .replace(/'/g, '&apos;');
    }

    /**
     * Parse XML-RPC response
     */
    parseXmlResponse(xmlText) {
        const parser = new DOMParser();
        const xmlDoc = parser.parseFromString(xmlText, "text/xml");

        // Check for fault
        const fault = xmlDoc.querySelector('methodResponse > fault');
        if (fault) {
            const errVal = this.parseXmlValue(fault.querySelector('value'));
            throw new Error(`Odoo Error: ${errVal.faultString} (${errVal.faultCode})`);
        }

        const param = xmlDoc.querySelector('methodResponse > params > param > value');
        if (!param) throw new Error('Invalid XML-RPC response');

        return this.parseXmlValue(param);
    }

    parseXmlValue(valNode) {
        const child = valNode.firstElementChild;
        if (!child) return valNode.textContent; // string fallback

        const type = child.tagName;

        if (type === 'string') return child.textContent;
        if (type === 'int' || type === 'i4') return parseInt(child.textContent, 10);
        if (type === 'double') return parseFloat(child.textContent);
        if (type === 'boolean') return child.textContent === '1';
        if (type === 'nil') return null;

        if (type === 'array') {
            const dataNode = child.querySelector('data');
            if (!dataNode) return [];
            return Array.from(dataNode.children).map(v => this.parseXmlValue(v));
        }

        if (type === 'struct') {
            const obj = {};
            Array.from(child.children).forEach(member => {
                const name = member.querySelector('name').textContent;
                const value = member.querySelector('value');
                obj[name] = this.parseXmlValue(value);
            });
            return obj;
        }

        return child.textContent;
    }

    /**
     * Execute XML-RPC fetch
     */
    async execute(endpoint, method, params) {
        const config = this.getConfig();
        const url = `${config.odooUrl}/xmlrpc/2/${endpoint}`;

        const body = this.buildXmlCall(method, params);

        try {
            const response = await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'text/xml'
                },
                body: body
            });

            if (!response.ok) {
                throw new Error(`HTTP Error: ${response.status} ${response.statusText}`);
            }

            const text = await response.text();
            return this.parseXmlResponse(text);

        } catch (error) {
            console.error('XML-RPC Call Failed:', error);
            throw error;
        }
    }

    /**
     * Authenticate and get UID
     */
    async authenticate() {
        const config = this.getConfig();

        // params: db, login, password, empty_dict
        const params = [
            config.database,
            config.username,
            config.password,
            {}
        ];

        console.log('Authenticating...');
        const uid = await this.execute('common', 'authenticate', params);

        if (!uid) {
            throw new Error('Authentication failed. Check credentials.');
        }

        this.uid = uid;
        console.log('Authenticated! UID:', uid);
        return uid;
    }

    /**
     * Search and Read (Main data fetching method)
     */
    async searchRead(model, domain = [], fields = []) {
        const config = this.getConfig();
        if (!this.uid) await this.authenticate();

        // execute_kw params: db, uid, password, model, method, args(list), kwargs(struct)
        const params = [
            config.database,
            this.uid,
            config.password,
            model,
            'search_read',
            [domain], // args: [domain]
            { fields: fields, limit: 100 } // kwargs
        ];

        return await this.execute('object', 'execute_kw', params);
    }

    /**
     * Helper: List Job Orders
     */
    async listJobOrders(searchTerm = '') {
        const config = this.getConfig();
        const model = config.modelJobOrder || 'project.project';

        let domain = [];
        if (searchTerm) {
            domain.push(['name', 'ilike', searchTerm]);
        }

        // We fetch name and id as minimal requirement
        return await this.searchRead(model, domain, ['id', 'name']);
    }

    /**
     * Helper: List Certificates for a Job Order
     */
    async listCertificates(jobOrderId) {
        const config = this.getConfig();
        const model = config.modelCertificate || 'quality.certificate';
        const relationField = config.fieldRelation || 'project_id';

        // Maps
        const fSerial = config.fieldSerial || 'slno';
        const fCertNo = config.fieldCertNo || 'certificate_number';
        const fIssue = config.fieldIssueDate || 'issue_date';
        const fExpiry = config.fieldExpiryDate || 'expiry_date';

        // Domain: Relation Field = jobOrderId
        const domain = [[relationField, '=', jobOrderId]];

        // Fields to fetch
        const fields = ['id', 'name', fSerial, fCertNo, fIssue, fExpiry];

        const results = await this.searchRead(model, domain, fields);

        // Normalize results
        return results.map(rec => ({
            id: rec.id,
            name: rec.name || 'No Name', // Default name

            // Normalized fields for our app
            serial: rec[fSerial] || '',
            certNo: rec[fCertNo] || '',
            issueDate: rec[fIssue] || '',
            expiryDate: rec[fExpiry] || ''
        }));
    }
}

// Export global
window.OdooClient = new OdooClient();

File: sticker_print_flutter/lib/services/odoo_client.dart
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart';
import '../models/app_config.dart';
import '../models/job_order.dart';
import '../models/certificate.dart';

/// Odoo XML-RPC Client - handles all communication with Odoo server
/// 
/// ✅ NO CORS ISSUES - Direct HTTP from Flutter app
class OdooClient {
  final AppConfig config;
  int? _uid;

  OdooClient(this.config);

  /// Build XML-RPC method call
  String _buildXmlCall(String methodName, List<dynamic> params) {
    final builder = XmlBuilder();
    builder.processing('xml', 'version="1.0"');
    builder.element('methodCall', nest: () {
      builder.element('methodName', nest: methodName);
      builder.element('params', nest: () {
        for (final param in params) {
          builder.element('param', nest: () {
            builder.element('value', nest: () {
              _buildXmlValue(builder, param);
            });
          });
        }
      });
    });
    return builder.buildDocument().toXmlString();
  }

  /// Convert Dart value to XML-RPC value
  void _buildXmlValue(XmlBuilder builder, dynamic value) {
    if (value == null) {
      builder.element('nil');
    } else if (value is bool) {
      builder.element('boolean', nest: value ? '1' : '0');
    } else if (value is int) {
      builder.element('int', nest: value.toString());
    } else if (value is double) {
      builder.element('double', nest: value.toString());
    } else if (value is String) {
      builder.element('string', nest: value);
    } else if (value is List) {
      builder.element('array', nest: () {
        builder.element('data', nest: () {
          for (final item in value) {
            builder.element('value', nest: () {
              _buildXmlValue(builder, item);
            });
          }
        });
      });
    } else if (value is Map) {
      builder.element('struct', nest: () {
        for (final entry in value.entries) {
          builder.element('member', nest: () {
            builder.element('name', nest: entry.key.toString());
            builder.element('value', nest: () {
              _buildXmlValue(builder, entry.value);
            });
          });
        }
      });
    }
  }

  /// Parse XML-RPC response
  dynamic _parseXmlResponse(String xmlText) {
    final document = XmlDocument.parse(xmlText);

    // Check for fault
    final fault = document.findAllElements('fault').firstOrNull;
    if (fault != null) {
      final faultValue = _parseXmlValue(fault.findElements('value').first);
      throw Exception(
          'Odoo Error: ${faultValue['faultString']} (${faultValue['faultCode']})');
    }

    // Parse response
    final param = document
        .findAllElements('methodResponse')
        .first
        .findAllElements('params')
        .first
        .findAllElements('param')
        .first
        .findElements('value')
        .first;

    return _parseXmlValue(param);
  }

  /// Parse XML value recursively
  dynamic _parseXmlValue(XmlElement valueNode) {
    final child = valueNode.childElements.firstOrNull;
    if (child == null) return valueNode.innerText;

    switch (child.name.local) {
      case 'string':
        return child.innerText;
      case 'int':
      case 'i4':
        return int.parse(child.innerText);
      case 'double':
        return double.parse(child.innerText);
      case 'boolean':
        return child.innerText == '1';
      case 'nil':
        return null;
      case 'array':
        final dataNode = child.findElements('data').first;
        return dataNode.findElements('value').map(_parseXmlValue).toList();
      case 'struct':
        final map = <String, dynamic>{};
        for (final member in child.findElements('member')) {
          final name = member.findElements('name').first.innerText;
          final value = member.findElements('value').first;
          map[name] = _parseXmlValue(value);
        }
        return map;
      default:
        return child.innerText;
    }
  }

  /// Execute XML-RPC call
  Future<dynamic> _execute(String endpoint, String method,
      List<dynamic> params) async {
    final url = '${config.odooUrl}/xmlrpc/2/$endpoint';
    final body = _buildXmlCall(method, params);

    final response = await http.post(
      Uri.parse(url),
      headers: {'Content-Type': 'text/xml'},
      body: body,
    );

    if (response.statusCode != 200) {
      throw Exception(
          'HTTP Error: ${response.statusCode} ${response.reasonPhrase}');
    }

    return _parseXmlResponse(response.body);
  }

  /// Authenticate and get UID
  Future<int> authenticate() async {
    final params = [
      config.database,
      config.username,
      config.password,
      {},
    ];

    final uid = await _execute('common', 'authenticate', params);

    if (uid == null || uid == false) {
      throw Exception('Authentication failed. Check credentials.');
    }

    _uid = uid as int;
    return _uid!;
  }

  /// Search and read records
  Future<List<dynamic>> searchRead(
    String model,
    List<dynamic> domain,
    List<String> fields,
  ) async {
    if (_uid == null) await authenticate();

    final params = [
      config.database,
      _uid,
      config.password,
      model,
      'search_read',
      [domain],
      {'fields': fields, 'limit': 100},
    ];

    final result = await _execute('object', 'execute_kw', params);
    return result as List<dynamic>;
  }

  /// List job orders
  Future<List<JobOrder>> listJobOrders({String searchTerm = ''}) async {
    final model = config.modelJobOrder;
    final domain = searchTerm.isEmpty
        ? []
        : [
            ['name', 'ilike', searchTerm]
          ];

    final results = await searchRead(model, domain, ['id', 'name']);
    return results.map((data) => JobOrder.fromOdoo(data as Map<String, dynamic>)).toList();
  }

  /// List certificates for a job order
  Future<List<Certificate>> listCertificates(int jobOrderId) async {
    final model = config.modelCertificate;
    final relationField = config.fieldRelation;

    final domain = [
      [relationField, '=', jobOrderId]
    ];

    final fields = [
      'id',
      'name',
      config.fieldSerial,
      config.fieldCertNo,
      config.fieldIssueDate,
      config.fieldExpiryDate,
    ];

    final results = await searchRead(model, domain, fields);
    return results
        .map((data) => Certificate.fromOdoo(
              data as Map<String, dynamic>,
              fieldSerial: config.fieldSerial,
              fieldCertNo: config.fieldCertNo,
              fieldIssueDate: config.fieldIssueDate,
              fieldExpiryDate: config.fieldExpiryDate,
            ))
        .toList();
  }
}
